client-v2/src/shared/util/shared.ts

```diff
@@ -26,38 +26,74 @@ const createGroup = (
 
 const getUUID = <T extends { uuid: string }>({ uuid }: T) => uuid;
 
-// this is horrible but is the only way to do it!
-const groupByIndex = (groups: Group[], index: number): Group | undefined =>
-  groups.find(g => parseInt(g.id || '0', 10) === index);
+const getGroupIndex = (id: string | null): number => parseInt(id || '0', 10);
 
 const getAllArticleFragments = (groups: Group[]) =>
   groups.reduce(
     (acc, { articleFragments }) => [...acc, ...articleFragments],
     [] as string[]
   );
 
-const addEmptyGroupsFromCollectionConfigForStage = (
+const addGroupsForStage = (
   groupIds: string[],
   entities: { [id: string]: Group },
   collectionConfig: CollectionConfig
 ) => {
   const groups = groupIds.map(id => entities[id]);
-  const addedGroups = collectionConfig.groups
-    ? collectionConfig.groups.map((name, i) => {
-        const existingGroup = groupByIndex(groups, i);
+  let groupsWithNames = groups.map(group => {
+    let name: string | null = null;
+    const groupNumberAsInt = getGroupIndex(group.id);
+    if (
+      collectionConfig.groups &&
+      groupNumberAsInt < collectionConfig.groups.length
+    ) {
+      name = collectionConfig.groups[groupNumberAsInt];
+    }
+    return { ...group, name };
+  });
+
+  // We may have empty groups in the config which would not show up in the normalised
+  // groups result. We need to add these into the groups array.
+  if (collectionConfig.groups) {
+    collectionConfig.groups.forEach((group, index) => {
+      if (
+        !groupsWithNames.some(addedGroup => {
```

Comment:
Could this live in a little helper that explains its purpose? We're a few indentations deep and it might make things a bit clearer -- but it might be just because it's half five and I should be on my bike already!

Reply:
Yes, excellent idea. I think it's much clearer now. Let me know what you think